lib/r10k/source/exec.rb

```diff
@@ -0,0 +1,43 @@
+require 'r10k/util/subprocess'
+require 'json'
+require 'yaml'
+
+class R10K::Source::Exec < R10K::Source::Hash
+  R10K::Source.register(:exec, self)
+
+  def initialize(name, basedir, options = {})
```

Comment:
All of the logic within this class is happening within the constructor, so you can't create an instance of this class without triggering side effects. Instead of using inheritance to use the `R10K::Source::Hash` class we can use composition, with something like the following:

```ruby
class R10K::Source::Exec < R10K::Source::Base
  def initialize(...)
    # ..
    @inner = null
  end

  def environments
    inner.environments
  end

  def desired_contents
    inner.desired_contents
  end
  
  private

  def inner
    envs = run_command_and_get_envs_hash
    @inner ||= R10K::Source::Hash.new(@name, @basedir, environments: envs)
  end
end
```

Thoughts?

Reply:
Pushed some new commits that should allow an instance of the class to be created without side-effects. Running of the configured command is now delayed until the source's `#environments` method is called.